### PR TITLE
test: resharding V3 and delayed receipts

### PR DESCRIPTION
**Pull request by Trisfald**
Adding a basic test-loop test for resharding v3 and delayed receipts splitting.

During the test I'm burning more gas than the chain capacity at every block height before resharding and one block after. Thus, in the resharding block height there are delayed receipts to be split.
A couple of blocks later I verify the transaction outcomes.

The test passes if the target account is not in the parent shard. It fails if the account belongs to either future child.

### CODE REVIEW

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 1]
```diff
@@ -113,11 +123,13 @@ struct TestReshardingParameters {
     shuffle_shard_assignment_for_chunk_producers: bool,
     track_all_shards: bool,
     /// Custom behavior executed at every iteration of test loop.
-    loop_action: Option<Box<dyn Fn(&mut TestLoopData, TestLoopDataHandle<ClientActorInner>)>>,
+    loop_action:
+        Option<Box<dyn Fn(&[TestData], &mut TestLoopData, TestLoopDataHandle<ClientActorInner>)>>,
```
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:127]: nit: Can you create a type alias for this / part of this?

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 2]
```diff
@@ -113,11 +123,13 @@ struct TestReshardingParameters {
...
     // When enabling shard shuffling with a short epoch length, sometimes a node might not finish
     // catching up by the end of the epoch, and then misses a chunk. This can be fixed by using a longer
     // epoch length, but it's good to also check what happens with shorter ones.
     all_chunks_expected: bool,
+    deploy_test_contract: Option<AccountId>,
```
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:132]: Can you add a comment please?

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 3]
```diff
@@ -268,6 +277,94 @@ fn fork_before_resharding_block(
...
+                        tracing::debug!(target: "test", ?tx_height, ?tx, ?status, "transaction status");
+                        assert_matches!(status, FinalExecutionStatus::SuccessValue(_));
+                    }
+                }
+            } else {
+                if next_block_has_new_shard_layout(client_actor.client.epoch_manager.clone(), &tip)
+                {
+                    tracing::debug!(target: "test", height=tip.height, "resharding height set");
+                    resharding_height.set(Some(tip.height));
+                }
```
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:325]: Can you also assert that there are delayed receipts at the post state of the resharding block?

**Review by wacban — COMMENTED** [review]
LGTM

It's an interesting setup where you have all of the test logic in one place and decide what to do based on height. By comparison in the older tests we typically split the test into multiple stages and at each stage we run the chain while using the logic for the current stage. I'm fine with either, just an observation.

**Review by wacban — APPROVED** [review]

File: integration-tests/src/test_loop/tests/resharding_v3.rs (hunk 3 above)
Comment (Trisfald) [reply integration-tests/src/test_loop/tests/resharding_v3.rs]: Good point, I've done a check using `congestion_info`.

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 4]
```diff
@@ -268,6 +277,94 @@ fn fork_before_resharding_block(
...
+    let resharding_height = Cell::new(None);
+    let nonce = Cell::new(102);
+    let txs = Cell::new(vec![]);
+    let latest_height = Cell::new(0);
+    let signer: Signer = create_user_test_signer(&signer_id).into();
+
+    Box::new(
+        move |node_datas: &[TestData],
+              test_loop_data: &mut TestLoopData,
+              client_handle: TestLoopDataHandle<ClientActorInner>| {
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:303]: I'll likely address this point in the second PR

**Pull request merged**